### ISSUE TITLE
merge() with a lambda results in wrong shape for model.fit()

### ISSUE THREAD

**Issue by carlthome** [fchollet/keras#2547]
When calling `model.fit(...)` [an error](https://github.com/fchollet/keras/blob/master/keras/engine/training.py#L100) occurs in `standardize_input_data()`:

> Exception: Error when checking model target: expected merge_2 to have shape (32, 1000, 1024) but got array with shape (352, 1000, 1024)

However, since simply adding `return arrays` before `# check shapes compatibility` and hoping the shapes are right, training works as expected so what's up? Quick guess: merge() should update `self.internal_input_shapes` or something.

```Python
batch_size = ...
timesteps = ...
features = ...
x, y = load_data()  # Both x and y have shape (n*batch_size, timesteps, features), where n : int > 0
input = Input(batch_shape=(batch_size, timesteps, features))
tdd1 = TimeDistributed(Dense(features))(input)
tdd2 = TimeDistributed(Dense(features))(input)
output1 = merge(
    inputs=[tdd1, tdd2],
    output_shape=(batch_size, timesteps, features),
    mode=lambda l: l[0])
output2 = merge(
    inputs=[tdd1, tdd2],
    output_shape=(batch_size, timesteps, features),
    mode=lambda l: l[1])
model = Model(input=[input], output=[output1, output2])
model.compile(...)
model.fit(x, y, batch_size)
```

**Comment by carlthome** [comment]
This seems related https://github.com/fchollet/keras/issues/2539 - guessing merge() returns an array but a list is expected or something.

**Comment by joelthchao** [comment]
@carlthome 
I reproduce your error and do some tracings into code. The error happens because you specify the `batch_size` in the argument `output_shape` in the `merge`. I modify them to `None` and everything goes well.
```python
input = Input(shape=(timesteps, features))
tdd1 = TimeDistributed(Dense(features))(input)
tdd2 = TimeDistributed(Dense(features))(input)
output1 = merge(
    inputs=[tdd1, tdd2],
    output_shape=(None, timesteps, features),
    mode=lambda l: l[0])
output2 = merge(
    inputs=[tdd1, tdd2],
    output_shape=(None, timesteps, features),
    mode=lambda l: l[1])

model = Model(input=[input], output=[output1, output2])
```
Can you justify if it works on your data?

**Comment by carlthome** [comment]
@joelthchao thanks for looking into this. 

Weirdly specifying the shape to variable batch size with `None` in the two `merge(...)`calls results in a NaN loss. Keeping the batch size fixed and not running `standardize_input_data(...)` results in proper loss values. The lambda function in both cases is the exact same function, so it supposedly returns tensors of the exact same dimensions.

Just for clarification this is the exact lambda I'm running:
```Python
lambda l: return l[2] * K.abs(l[0]) / (K.abs(l[0]) + K.abs(l[1]))
```

**Comment by joelthchao** [comment]
@carlthome Can you try `K.abs(l[0]) / (K.abs(l[0]) + K.abs(l[1]) + 1e-7)` to remove the possibility of zero  division?

**Comment by carlthome** [comment]
Wow, good find. Silly me. Thanks! :smile:
